pkg/frontend/querymiddleware/astmapper/pruning_new_test.go

```diff
@@ -0,0 +1,179 @@
+)
+
+func TestQueryPrunerV2(t *testing.T) {
+	pruner := NewQueryPrunerV2(context.Background(), log.NewNopLogger())
+
+	for _, tt := range []struct {
+		in  string
+		out string
+	}{
+		// Non-prunable expressions.
+		{
+			`foo`,
+			`foo`,
+		},
+		{
+			`foo[1m]`,
+			`foo[1m]`,
+		},
+		{
+			`foo{bar="baz"}[1m]`,
+			`foo{bar="baz"}[1m]`,
+		},
+		{
+			`foo{bar="baz"}`,
+			`foo{bar="baz"}`,
+		},
+		{
+			`quantile(0.9,foo)`,
+			`quantile(0.9,foo)`,
+		},
+		{
+			`histogram_quantile(0.5, rate(bar1{baz="blip"}[30s]))`,
+			`histogram_quantile(0.5, rate(bar1{baz="blip"}[30s]))`,
+		},
+		{
+			`count(up)`,
+			`count(up)`,
+		},
+		{
+			`avg(rate(foo[1m]))`,
+			`avg(rate(foo[1m]))`,
+		},
+		{
+			`vector(0)`,
+			`vector(0)`,
+		},
+		{
+			`up < -Inf`,
+			`up < -Inf`,
+		},
+		{
+			`avg(rate(foo[1m])) < (-Inf)`,
+			`avg(rate(foo[1m])) < (-Inf)`,
+		},
+		{
+			`Inf * -1`,
+			`Inf * -1`,
+		},
+		{
+			`avg(rate(foo[1m])) < (-1 * +Inf)`,
+			`avg(rate(foo[1m])) < (-1 * +Inf)`,
+		},
+		{
+			`(-1 * -Inf) < avg(rate(foo[1m]))`,
+			`(-1 * -Inf) < avg(rate(foo[1m]))`,
+		},
+		{
+			`vector(0) < -Inf or avg(rate(foo[1m]))`,
+			`vector(0) < -Inf or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) or vector(0) < -Inf`,
+			`avg(rate(foo[1m])) or vector(0) < -Inf`,
+		},
+		{
+			`avg(rate(foo[1m])) or (vector(0) < -Inf)`,
+			`avg(rate(foo[1m])) or (vector(0) < -Inf)`,
+		},
+		{
+			`((-1 * -Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((-1 * -Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`(avg(rate(foo[2m])) < (+1 * -Inf)) or avg(rate(foo[1m]))`,
+			`(avg(rate(foo[2m])) < (+1 * -Inf)) or avg(rate(foo[1m]))`,
+		},
+		{
+			`((-1 * -Inf) == avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((-1 * -Inf) == avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) unless ((-1 * -Inf) < avg(rate(foo[2m])))`,
+			`avg(rate(foo[1m])) unless ((-1 * -Inf) < avg(rate(foo[2m])))`,
+		},
+		{
+			`((2 * +Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((2 * +Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) or avg(rate(bar[1m]))`,
+			`avg(rate(foo[1m])) or avg(rate(bar[1m]))`,
+		},
+		{ // The const expression is on the wrong side.
+			`(vector(0) == 1) and on() (avg(rate(foo[1m])))`,
+			`(vector(0) == 1) and on() (avg(rate(foo[1m])))`,
+		},
+		{ // Matching on labels.
+			`(avg(rate(foo[1m]))) and on(id) (vector(0) == 1)`,
+			`(avg(rate(foo[1m]))) and on(id) (vector(0) == 1)`,
```

Comment:
Couldn't this be simplified to `vector(0) == 1`? The right side will never produce a result, so we'll never return anything from the left side.

Reply:
out of scope